### ISSUE TITLE
Error: The following cols are missing from `new_data` after #865

### ISSUE THREAD

**Issue by hsbadr** [tidymodels/recipes#1008]
## The problem

When creating a recipe from the data (not formula; using `update_role()`) and adding it to a workflow with a model overriding the formula (using `workflows::add_model(spec, formula = <formula>)`), `recipes` complains about missing cols in the `new_data` that are actually not missing and not used in the `<formula>`. The example bellow works fine before #865.

## Reproducible example

``` r
library(reprex)
library(parsnip)
library(recipes)
library(workflows)

data(ad_data, package = "modeldata")

ad_rec <-
  ad_data %>%
  recipe() %>%
  update_role(Class, new_role = "outcome") %>%
  update_role(tau, VEGF, new_role = "predictor") %>%
  step_normalize(all_numeric_predictors())

ad_spec <-
  logistic_reg() %>%
  set_engine("glm")

ad_wf <-
  workflow() %>%
  add_recipe(ad_rec) %>%
  add_model(ad_spec, formula = Class ~ tau + VEGF) %>%
  fit(data = ad_data)

predict(ad_wf, ad_data[1:4,])
#> Error:
#> ! The following cols are missing from `new_data`: ACE_CD143_Angiotensin_....

#> Backtrace:
#>      ▆
#>   1. ├─stats::predict(ad_wf, ad_data[1:4, ])
#>   2. └─workflows:::predict.workflow(ad_wf, ad_data[1:4, ])
#>   3.   └─workflows:::forge_predictors(new_data, workflow) at workflows/R/predict.R:62:2
#>   4.     ├─hardhat::forge(new_data, blueprint = mold$blueprint) at workflows/R/predict.R:69:2
#>   5.     └─hardhat:::forge.data.frame(new_data, blueprint = mold$blueprint) at hardhat/R/forge.R:68:2
#>   6.       ├─hardhat::run_forge(blueprint, new_data = new_data, outcomes = outcomes) at hardhat/R/forge.R:81:2
#>   7.       └─hardhat:::run_forge.default_recipe_blueprint(...) at hardhat/R/forge.R:135:2
#>   8.         └─hardhat:::forge_recipe_default_process(...) at hardhat/R/blueprint-recipe-default.R:357:2
#>   9.           ├─recipes::bake(object = rec, new_data = new_data) at hardhat/R/blueprint-recipe-default.R:440:2
#>  10.           └─recipes:::bake.recipe(object = rec, new_data = new_data) at recipes/R/recipe.R:500:2
#>  11.             └─recipes:::check_new_data_columns(object, new_data) at recipes/R/recipe.R:568:2
#>  12.               └─rlang::abort(msg, call = NULL) at recipes/R/misc.R:744:6
```

<sup>Created on 2022-06-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by juliasilge** [comment]
If you don't pass a special model formula to `add_model()` it does work:

``` r
library(parsnip)
library(recipes)
#> Loading required package: dplyr
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
library(workflows)

data(ad_data, package = "modeldata")

ad_rec <-
  ad_data %>%
  recipe() %>%
  update_role(Class, new_role = "outcome") %>%
  update_role(tau, VEGF, new_role = "predictor") %>%
  step_normalize(all_numeric_predictors())

ad_spec <-
  logistic_reg() %>%
  set_engine("glm")

ad_wf <-
  workflow() %>%
  add_recipe(ad_rec) %>%
  add_model(ad_spec) %>%
  fit(data = ad_data)

predict(ad_wf, ad_data[1:4,])
#> # A tibble: 4 × 1
#>   .pred_class
#>   <fct>      
#> 1 Control    
#> 2 Impaired   
#> 3 Control    
#> 4 Control
```

<sup>Created on 2022-06-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I'm not sure we ever expected what you showed to be a typical pattern; can you tell us more about your use case? Are you using a model that needs the special formula like [multilevelmod](https://multilevelmod.tidymodels.org/)?

**Comment by hsbadr** [comment]
Thanks @juliasilge!

> If you don't pass a special model formula to `add_model()` it does work:

Yes, and if I add a formula in the recipe, it will work too.

> I'm not sure we ever expected what you showed to be a typical pattern; can you tell us more about your use case? Are you using a model that needs the special formula like [multilevelmod](https://multilevelmod.tidymodels.org/)?

Yes, [this vignette](https://hsbadr.github.io/additive/articles/GetStarted.html) now fails and it shouldn't. In general, I need to pass a special formula for some additive and bayesian models, and could even need to use a `formula.override` to pass a list of formulae (I can provide examples, if needed -- but it's out of the scope of this issue).

I'd expect that `recipes` won't fail due to missing cols that are already provided in the `new_data` (even though they may not be used). I guess that `workflows`/`hardhat` does remove the unused variables by the provided formula, and so, `recipes` would need to only check the required variables.

**Comment by topepo** [comment]
With the new recipes (that I'm about to send to CRAN) and the new workflows (happens immediately after), you would declare that those columns with unspecified roles are special cases and should required at bake-time. 

``` r
library(reprex)
library(parsnip)
library(recipes)
#> Loading required package: dplyr
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
library(workflows)
library(hardhat)

data(ad_data, package = "modeldata")

cols_used <- c("Class", "tau", "VEGF")

ad_rec <-
  ad_data %>%
  recipe() %>%
  update_role(Class, new_role = "outcome") %>%
  update_role(all_of(cols_used), new_role = "predictor") %>%
  update_role(-Class, -tau, -VEGF, new_role = "other") %>% 
  step_normalize(all_numeric_predictors())

ad_spec <-
  logistic_reg() %>%
  set_engine("glm")

custom_bp <- default_recipe_blueprint(bake_dependent_roles = "other")

ad_wf <-
  workflow() %>%
  add_recipe(ad_rec, blueprint = custom_bp) %>%
  add_model(ad_spec, formula = Class ~ tau + VEGF) %>%
  fit(data = ad_data)

predict(ad_wf, ad_data[1:4,])
#> # A tibble: 4 × 1
#>   .pred_class
#>   <fct>      
#> 1 Control    
#> 2 Impaired   
#> 3 Control    
#> 4 Control
```

<sup>Created on 2022-06-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by hsbadr** [comment]
Thanks @topepo! It's great that we can specify variables required at bake-time, even if they have no role. Though, I still think that the logic in #1009 is the expected behavior. The old filter based on `is.na(role) | !(role %in% exclusions)` just applies the explicit exclusions and let `NA` roles pass while `!is.na(role) & !(role %in% exclusions)` does filter them as well (they are literally not required). As stated in the news, #865 is intended to let `bake()` error if `new_data` doesn't contain all **the required columns**. In the OP example, the `new_data` includes all columns and complaining about columns with `NA` is unnecessary.

**Comment by hsbadr** [comment]
Fixed in #1011.
